### ISSUE TITLE
typescript definitions

### ISSUE THREAD

**Issue by danprince** [smallwins/slack#89]
Is there an interest in offering TypeScript definitions for this project?

We currently maintain local types for the promise style methods for the subset of APIs that we're using, and for now it's easiest for them to keep them local, but if there are others out there who were interested in helping keep them up to date, we could start contributing them to this repository or a new `@types/slack` package.

Could also look at the viability of generating a declaration file from the existing `api.json` format. That would be helpful for intellisense, but I'm not sure we can reliably infer anything useful about the response types from the docs.

**Comment by brianleroux** [comment]
Yes! I've been meaning to try and do this for an age. Help here from someone who knows what they're doing would be really great. (I'm not up to full speed on TS defs.)

Definitely would be best to generate them from `api.json`. Alas but afaict the only thing responses have in common is they're always an `Object`. 🤷‍♂️

**Comment by danprince** [comment]
Here's a sketch of what I think we could generate from `api.json`:

```ts
declare module "slack" {
  export let auth: {
    test({ token }): Promise<any>;
    test({ token }, callback): void;
  }

  // ...

  export let chat: {
    postMessage({ token, channel, text, user, as_user, attachments, link_names, parse }): Promise<any>;
    postMessage({ token, channel, text, user, as_user, attachments, link_names, parse }, callback): void;
  }
}
```

That's enough info to tell TypeScript users which methods exist and which params they accept, unfortunately it doesn't get at most of the value of having types, as the responses and the individual params all get the `any` type. Still, better than nothing.

For comparison, our handwritten definitions look something like this:

```ts
namespace Slack.Auth {
  export interface Params {
    token: string;
  }

  export interface Response {
    ok: boolean;
    url: string;
    team: string;
    user: string;
    team_id: string;
    user_id: string;
  }
}

declare module "slack" {
  export let auth: {
    test(params: Slack.Auth.Params): Promise<Slack.Auth.Response>;
  }
}
```

It'd be a much bigger job to get the whole API covered like this, but it hugely improves the experience for TypeScript consumers.

**Comment by brianleroux** [comment]
🆗🆒  this is really good stuff. lots of authortime wins. tooling support is huge too. really love how readable these defs are. And I think either could be generated but I like starting with the simple case first. 

Also! Know the Slack team is working towards something here. cc @Roach 🙋‍♂️

**Comment by danprince** [comment]
Took a bit of work, but finally got the generated definitions type checking against our codebase! Published them in 9.0.9.

Going to follow up with further work in #94 and keep thinking about ways we could reduce the generated file size.
